include/GafferUI/AnimationGadget.h

```diff
@@ -0,0 +1,189 @@
+//
+//  Copyright (c) 2018, Matti Gruener. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERUI_ANIMATIONGADGET_H
+#define GAFFERUI_ANIMATIONGADGET_H
+
+#include "GafferUI/Gadget.h"
+#include "GafferUI/ViewportGadget.h"
+
+#include "Gaffer/Animation.h"
+#include "Gaffer/ScriptNode.h"
+
+namespace GafferUI
+{
+
+/// Aliases that define the intended use of each
+/// Gadget::Layer by the AnimationGadget components.
+namespace AnimationLayer
+{
+	constexpr Gadget::Layer Grid = Gadget::Layer::Back;
+	constexpr Gadget::Layer Curves = Gadget::Layer::MidBack;
+	constexpr Gadget::Layer Keys = Gadget::Layer::Main;
+	constexpr Gadget::Layer Axes = Gadget::Layer::MidFront;
+	constexpr Gadget::Layer Overlay = Gadget::Layer::Front;
+};
+
+class GAFFERUI_API AnimationGadget : public Gadget
+{
+
+public :
+
+	typedef boost::signal<void (Gaffer::Plug *, bool)> BinarySignal;
+
+	AnimationGadget( Gaffer::ScriptNode *scriptNode );
+
+	~AnimationGadget() override;
+
+	IE_CORE_DECLARERUNTIMETYPEDEXTENSION( GafferUI::AnimationGadget, AnimationGadgetTypeId, Gadget );
+
+	void setVisiblePlugs( const std::vector<Gaffer::Plug *> &plugs );
+	void setEditablePlugs( const std::vector<Gaffer::Plug *> &plugs );
+	void setFrame( float frame );
```

Comment:
Although we don't currently expose it to users, the API is such that different editors can use different contexts (see `Editor.setContext()`). Really the AnimationGadget needs a setContext/getContext pair rather than a frame for this to continue to work. Internally we shouldn't be manipulating the ScriptNode context directly...

Reply:
Addressed this one, too. I hope it's better now.